app/src/main/java/ch/epfl/sdp/drone3d/service/impl/drone/DroneExecutorImpl.kt

```diff
@@ -25,135 +26,175 @@ import kotlin.math.roundToInt
         val future = completable
-                .andThen(instance.mission.pauseMission())
-                .andThen(instance.mission.clearMission())
-                .andThen(instance.action.returnToLaunch())
+            .andThen(instance.mission.pauseMission())
+            .andThen(instance.mission.clearMission())
+            .andThen(instance.action.returnToLaunch())
 
         // Go to location
         data.addSubscription(
-                future.subscribe(
-                        {
-                            data.getMutableMission().postValue(listOf(DroneUtils.generateMissionItem(returnLocation.latitudeDeg, returnLocation.longitudeDeg, returnLocation.absoluteAltitudeM)))
-                            ToastHandler.showToastAsync(context, R.string.drone_home_success)
-                        },
-                        {
-                            data.getMutableMission().postValue(null)
-                            ToastHandler.showToastAsync(context, R.string.drone_home_error)
-                        }
-                )
+            future.subscribe(
+                {
+                    data.getMutableMission().postValue(
+                        listOf(
+                            DroneUtils.generateMissionItem(
+                                returnLocation.latitudeDeg,
+                                returnLocation.longitudeDeg,
+                                returnLocation.absoluteAltitudeM
+                            )
+                        )
+                    )
+                    ToastHandler.showToastAsync(context, R.string.drone_home_success)
+                },
+                {
+                    data.getMutableMission().postValue(null)
+                    ToastHandler.showToastAsync(context, R.string.drone_home_error)
+                }
+            )
         )
 
         // Land when arrived
         data.addSubscription(
-                instance.telemetry.position.subscribe(
-                        { pos ->
-                            val isRightPos = distance(pos, returnLocation) == 0
-                            val isStopped = data.getSpeed().value?.roundToInt() == 0
-                            if (isRightPos.and(isStopped)) instance.action.land().blockingAwait(1, TimeUnit.SECONDS)
-                            data.getMutableMissionPaused().postValue(true)
-                        },
-                        { e -> Timber.e("ERROR LANDING : $e") }
-                )
+            instance.telemetry.position.subscribe(
+                { pos ->
+                    performLanding(instance, pos, returnLocation)
+                },
+                { e -> Timber.e("ERROR LANDING : $e") }
+            )
         )
     }
+
+    private fun performLanding(
+        instance: System,
+        pos: Telemetry.Position,
+        returnLocation: Telemetry.Position
+    ) {
+        val isRightPos = distance(pos, returnLocation) == 0
+        val isStopped = data.getSpeed().value?.roundToInt() == 0
```

Comment:
Maybe veriying that the speed() is under a certain float treshold, rounding to Int seems arbitrary. 
If speed is in m/s, 0.5 meters per second might still be fast